core/src/main/scala/kafka/log/UnifiedLog.scala

```diff
@@ -770,8 +770,8 @@ class UnifiedLog(@volatile var logStartOffset: Long,
 
     val appendInfo = analyzeAndValidateRecords(records, origin, ignoreRecordSize, !validateAndAssignOffsets, leaderEpoch)
 
-    // return if we have no valid messages or if this is a duplicate of the last appended entry
-    if (appendInfo.validBytes <= 0) appendInfo
+    // return if we have no valid messages
+    if (!appendInfo.hasValidBytes) appendInfo
```

Comment:
I think this may capture the intent a bit better. The comment around duplicates of last appended entries seems to be a stale comment that was never updated. Appends don't get deduped in `analyzeAndValidateRecords`.